In [8]:
%cd "C:\Users\loren\Documents\COMP3608-Project"
import pandas as pd

#Algorithms used
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC

#Scalers to test
from sklearn.discriminant_analysis import StandardScaler
from sklearn.preprocessing import MinMaxScaler

#Sampling techniques to test
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler

from HelpFunctions.comparison_pipline import model_comparison

C:\Users\loren\Documents\COMP3608-Project


c:\Python312\Lib\site-packages\IPython\core\magics\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [9]:
#Model, resampling techniques and sampling methods to be compared are all added to a dict
models_to_test = {
    'Logistic Regression': LogisticRegression(random_state=42, max_iter=1000, n_jobs=-1),
    'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
    'Linear SVC': LinearSVC(dual=False, random_state=42, max_iter=2000)
}


In [10]:

scalers = {
        'None': None,
        'StandardScaler': StandardScaler(),
        'MinMaxScaler': MinMaxScaler()
}


## Data Scaling Techniques

Feature scaling is a crucial preprocessing step for many machine learning algorithms, especially those that are sensitive to the magnitude of input features. Without scaling, features with larger values might disproportionately influence the model, leading to suboptimal performance. We will explore the following scaling techniques:

---

### 1. `None`: No Scaler (Baseline)

*   **Purpose:**
    This approach serves as a **baseline** to understand the impact of feature scaling. By training models on unscaled data, we can observe if and how much the chosen algorithms are affected by differences in feature magnitudes. This helps to justify the use of scaling if scaled versions perform better.

*   **Applicability & Considerations:**
    *   **Tree-based models** (like Random Forest and Decision Trees) are generally insensitive to feature scaling because their splitting decisions are based on single features and thresholds, not on distances or magnitudes across features. For these models, `None` might yield similar performance to scaled versions.
    *   **Distance-based algorithms** (like K-Nearest Neighbors, SVMs) and **gradient-based algorithms** (like Logistic Regression, Neural Networks, or anything optimized with gradient descent) are highly sensitive to feature scaling. Without scaling, features with larger values and wider ranges can dominate the distance calculations or the gradient updates, leading to slower convergence or a suboptimal solution.
    *   For algorithms like **Linear SVM (LinearSVC)** and **Logistic Regression**, using unscaled data when features have vastly different ranges is generally not recommended.

---

### 2. `StandardScaler`: Standardization

*   **How it Works:**
    `StandardScaler` transforms the data by removing the mean and scaling to unit variance. For each feature $x$, the standardized value $x_{scaled}$ is calculated as:
    $$
    x_{scaled} = \frac{x - \mu}{\sigma}
    $$
    where:
    *   $\mu$ is the mean of the feature in the training set.
    *   $\sigma$ is the standard deviation of the feature in the training set.

    The result is that each feature will have a mean of 0 and a standard deviation of 1 after transformation.

*   **Purpose:**
    To center the data around zero and give all features a similar scale (unit variance). This is particularly important for algorithms that assume data is centered around zero or that features have similar variance, such as:
    *   **Support Vector Machines (SVMs / LinearSVC):** The optimization process for finding the maximal margin hyperplane is sensitive to feature scales.
    *   **Logistic Regression (and other models using gradient descent):** Standardization can help gradient descent converge faster and more reliably.
    *   Principal Component Analysis (PCA).

*   **Advantages:**
    *   Maintains the shape of the original distribution (does not make it Gaussian if it wasn't).
    *   Less affected by outliers compared to `MinMaxScaler` because it uses mean and standard deviation, which are influenced by outliers, but the scaling itself doesn't compress values into a strict range determined by min/max.
    *   Often a good default choice for many algorithms.

*   **Disadvantages:**
    *   The transformed features do not have a strictly bounded range (e.g., between 0 and 1), which might be desirable for some specific algorithms or interpretations, though less common.

---

### 3. `MinMaxScaler`: Normalization (Min-Max Scaling)

*   **How it Works:**
    `MinMaxScaler` transforms features by scaling each feature to a given range, typically between 0 and 1 (which is the default). For each feature $x$, the scaled value $x_{scaled}$ is calculated as:
    $$
    x_{scaled} = \frac{x - \min(x)}{\max(x) - \min(x)}
    $$
    where:
    *   $\min(x)$ is the minimum value of the feature in the training set.
    *   $\max(x)$ is the maximum value of the feature in the training set.

    If a specific range `(a, b)` is desired instead of `(0, 1)`, the formula becomes:
    $$
    x_{scaled} = \frac{x - \min(x)}{\max(x) - \min(x)} \times (b - a) + a
    $$

*   **Purpose:**
    To scale all features into a fixed range, usually [0, 1]. This can be beneficial for algorithms that:
    *   Expect inputs within a specific bounded range (e.g., some neural network activation functions historically preferred inputs in [0,1] or [-1,1]).
    *   Are sensitive to feature magnitudes, similar to `StandardScaler`. For algorithms like SVMs and Logistic Regression, `MinMaxScaler` can also be effective.

*   **Advantages:**
    *   Guarantees that all features will have the exact same scale (e.g., [0, 1]), which can be useful if strict bounds are required.
    *   Can be good when the data distribution is not Gaussian and the algorithm does not assume any specific distribution.

*   **Disadvantages:**
    *   **Highly sensitive to outliers:** If there are very large or very small outliers in a feature, they will become the new `min(x)` or `max(x)`. This can cause the majority of the "normal" data points to be compressed into a very small sub-interval of the [0, 1] range, potentially diminishing the variance and discriminative power of that feature for the bulk of the data.
    *   May compress the data too much if the original standard deviation is small, potentially losing some information about variance.

In [11]:

samplers = {
    'None': None,
    'SMOTE': SMOTE(random_state=42), 
    'RUS': RandomUnderSampler(random_state=42), 
    'ROS': RandomOverSampler(random_state=42)   
}

## Data Sampling Techniques

From the EDAs done we saw that credit card fraud is highly imbalanced for dataset 2 and dataset 3, with a very small percentage of transactions being fraudulent (the minority class) compared to legitimate ones (the majority class). Training models directly on such imbalanced data can lead to classifiers that are biased towards the majority class, performing poorly in identifying fraud. To address this, the following sampling techniques will be explored

---

### 1. `None`: No Sampling (Baseline)

*   **Purpose:**
    This approach serves as a **baseline**. By training models on the raw, imbalanced data, we can establish a performance benchmark. The results from this baseline will help quantify the actual impact and benefit of applying different sampling techniques.

*   **Advantages:**
    *   Represents the true underlying distribution of the data.
    *   No artificial data is introduced, and no original data points are discarded at the sampling stage.
    *   Allows evaluation of how well a model can inherently handle imbalance.

*   **Disadvantages:**
    *   Models are likely to be biased towards the majority class (legitimate transactions).
    *   May result in poor recall for the minority class (fraudulent transactions), meaning many fraud cases might be missed.
    *   Metrics like accuracy can be misleading, as a model predicting everything as "legitimate" would achieve high accuracy.

---

### 2. `SMOTE`: Synthetic Minority Over-sampling Technique

*   **How it Works:**
    SMOTE is an over-sampling technique that creates **synthetic** samples for the minority class (fraudulent transactions) rather than just duplicating existing ones. For each minority class sample:
    1.  It finds its *k*-nearest neighbors (also from the minority class).
    2.  It randomly selects one or more of these neighbors.
    3.  New synthetic samples are generated along the line segment joining the original minority sample and its selected neighbor(s) in the feature space.
    The `SMOTE(random_state=42)` ensures that the random choices made during this process (like selecting neighbors or the point along the line segment) are reproducible.

*   **Purpose:**
    To increase the representation of the minority class by generating new, plausible minority samples, thereby helping to balance the class distribution and provide more "examples" for the model to learn the characteristics of the minority class.

*   **Advantages:**
    *   Can lead to better generalization for the minority class by creating a more diverse set of minority samples compared to simple over-sampling.
    *   Helps to make the decision regions for the minority class less specific to the exact original minority samples, potentially reducing overfitting that ROS might cause.
    *   Provides more robust "signals" for identifying fraudulent patterns.

*   **Disadvantages:**
    *   Can create noisy samples if the original minority samples are themselves noisy or if the synthetic samples are generated in regions that overlap significantly with the majority class, potentially blurring decision boundaries.
    *   Does not consider the majority class when generating samples, which might lead to the creation of synthetic samples in areas of high majority class density.
    *   Can be computationally more intensive than simpler sampling methods.

---

### 3. `RUS`: Random Under-Sampler

*   **How it Works:**
    RUS aims to balance class distribution by randomly removing samples from the **majority class** (legitimate transactions). It continues to discard majority class samples until the desired ratio between minority and majority class instances is achieved (often aiming for a 1:1 ratio, or a predefined sampling strategy).
    The `RandomUnderSampler(random_state=42)` ensures that the random selection of majority class samples to be discarded is reproducible.

*   **Purpose:**
    To reduce the skewness in the dataset by decreasing the number of majority class samples, making the dataset more balanced and potentially reducing the computational burden of training.

*   **Advantages:**
    *   Can significantly reduce the size of the training dataset, leading to faster model training times.
    *   Can help prevent models from being overwhelmed by the sheer volume of legitimate transactions, allowing them to pay more "attention" to the fraudulent ones.

*   **Disadvantages:**
    *   **Potential loss of important information:** By randomly discarding majority class samples, we might remove legitimate transactions that are crucial for defining the decision boundary between fraudulent and non-fraudulent behavior (e.g., legitimate transactions that look somewhat similar to fraudulent ones).
    *   May not be suitable if the original dataset size is small, as further reducing it could lead to insufficient data for robust model training.
    *   The random nature means that different runs (without a fixed `random_state`) could lead to different subsets and potentially different model performance.

---

### 4. `ROS`: Random Over-Sampler

*   **How it Works:**
    ROS aims to balance class distribution by randomly duplicating samples from the **minority class** (fraudulent transactions). Existing minority class samples are selected at random (with replacement) and added to the dataset until the desired ratio between minority and majority class instances is achieved.
    The `RandomOverSampler(random_state=42)` ensures that the random selection of minority class samples for duplication is reproducible.

*   **Purpose:**
    To increase the representation of the minority class by increasing its sample size, thereby balancing the dataset.

*   **Advantages:**
    *   Simple to understand and implement.
    *   No information from the original dataset is lost (unlike RUS).
    *   Can sometimes be effective for algorithms that are sensitive to class distribution.

*   **Disadvantages:**
    *   **Prone to overfitting:** Since it merely makes exact copies of existing minority samples, the model might learn these specific instances too well without generalizing to new, unseen fraudulent transactions. The model might become too specific to the duplicated patterns.
    *   Does not add any genuinely "new" information or variability to the minority class.
    *   Can significantly increase the size of the training dataset, potentially increasing training time, though usually less of a concern than the overfitting risk for fraud data.


In [12]:
#dataset3
df = pd.read_csv('card_transdata.csv', index_col=0)
binary_columns = {'repeat_retailer': 'bool', 'used_chip': 'bool',
                  'used_pin_number': 'bool', 'online_order': 'bool', 'fraud': 'bool'}
df = df.astype(binary_columns)


In [13]:

# Define features and target
X = df.drop(columns=["fraud"])
y = df["fraud"]

numerical_cols_actual = [col for col in X.columns if X[col].dtype in ['float64', 'int64', 'float32', 'int32']]
print(f"Using numerical columns for scaling: {numerical_cols_actual}")


Using numerical columns for scaling: ['distance_from_last_transaction', 'ratio_to_median_purchase_price']


In [14]:
# Run the pipeline
all_results = model_comparison(models_to_test, scalers, samplers, X, y, numerical_cols_actual, verbose=1)

Splitting data: test_size=0.2, random_state=42, stratify=y
Train shape: (800000, 6), Test shape: (200000, 6)
Test label distribution:
fraud
False    0.912595
True     0.087405
Name: proportion, dtype: float64


=== Evaluating Model: Logistic Regression ===

=== Evaluating Model: Random Forest ===

=== Evaluating Model: Linear SVC ===
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for RO

In [15]:
all_results

Model          Scaler Sampler  Accuracy  Precision  \
0         Random Forest    MinMaxScaler    None  0.981155   0.997749   
1         Random Forest            None    None  0.980790   0.992347   
2         Random Forest  StandardScaler    None  0.980780   0.991849   
3         Random Forest    MinMaxScaler     ROS  0.978430   0.957156   
4         Random Forest            None     ROS  0.976125   0.925976   
5         Random Forest  StandardScaler     ROS  0.976010   0.924607   
6         Random Forest    MinMaxScaler   SMOTE  0.952125   0.694155   
7         Random Forest            None   SMOTE  0.931895   0.577604   
8         Random Forest  StandardScaler   SMOTE  0.931165   0.574144   
9   Logistic Regression            None    None  0.954775   0.921042   
10  Logistic Regression  StandardScaler    None  0.954755   0.920927   
11        Random Forest    MinMaxScaler     RUS  0.913450   0.502949   
12        Random Forest  StandardScaler     RUS  0.901355   0.464576   
13        Random Forest            None     RUS  0.901045   0.463627   
14           Linear SVC  StandardScaler     ROS  0.886700   0.423494   
15           Linear SVC            None     ROS  0.886695   0.423481   
16           Linear SVC  StandardScaler   SMOTE  0.886530   0.423046   
17           Linear SVC            None   SMOTE  0.886455   0.422768   
18           Linear SVC  StandardScaler     RUS  0.886045   0.421755   
19           Linear SVC            None     RUS  0.886045   0.421751   
20           Linear SVC    MinMaxScaler     ROS  0.885770   0.421063   
21           Linear SVC    MinMaxScaler   SMOTE  0.885620   0.420701   
22  Logistic Regression            None     RUS  0.882435   0.413735   
23  Logistic Regression  StandardScaler     RUS  0.882400   0.413657   
24  Logistic Regression            None     ROS  0.881915   0.412558   
25  Logistic Regression  StandardScaler     ROS  0.881850   0.412400   
26  Logistic Regression  StandardScaler   SMOTE  0.881790   0.412274   
27  Logistic Regression            None   SMOTE  0.881775   0.412159   
28           Linear SVC    MinMaxScaler     RUS  0.878735   0.403665   
29  Logistic Regression    MinMaxScaler    None  0.942765   0.926733   
30  Logistic Regression    MinMaxScaler     ROS  0.875065   0.395617   
31  Logistic Regression    MinMaxScaler   SMOTE  0.874865   0.395194   
32           Linear SVC  StandardScaler    None  0.938895   0.911066   
33           Linear SVC            None    None  0.938895   0.911066   
34           Linear SVC    MinMaxScaler    None  0.938065   0.909749   
35  Logistic Regression    MinMaxScaler     RUS  0.842600   0.335372   

      Recall  F1 Score   ROC AUC  
0   0.786168  0.879411  0.958454  
1   0.786282  0.877378  0.946634  
2   0.786568  0.877361  0.945886  
3   0.788513  0.864689  0.958992  
4   0.790001  0.852601  0.947014  
5   0.789943  0.851987  0.947308  
6   0.808478  0.746968  0.962865  
7   0.821749  0.678378  0.961545  
8   0.822607  0.676276  0.962301  
9   0.527830  0.671079  0.928591  
10  0.527659  0.670910  0.928608  
11  0.834048  0.627502  0.963127  
12  0.843258  0.599094  0.963100  
13  0.842172  0.598030  0.962914  
14  0.819976  0.558526  0.937437  
15  0.819976  0.558515  0.937440  
16  0.819690  0.558070  0.937440  
17  0.818546  0.557562  0.937068  
18  0.818660  0.556707  0.937122  
19  0.818603  0.556690  0.937123  
20  0.818546  0.556078  0.937128  
21  0.818660  0.555789  0.937150  
22  0.827470  0.551647  0.938846  
23  0.827527  0.551590  0.938859  
24  0.828042  0.550726  0.938884  
25  0.827985  0.550572  0.938878  
26  0.828156  0.550498  0.938881  
27  0.827241  0.550193  0.938625  
28  0.811624  0.539170  0.934627  
29  0.374807  0.533746  0.918991  
30  0.813683  0.532385  0.934880  
31  0.813855  0.532039  0.934895  
32  0.333448  0.488211  0.924636  
33  0.333448  0.488211  0.924636  
34  0.323494  0.477276  0.924119  
35  0.815686  0.475316  0.924706

Based on the results for dataset 3, `Random Forest` with no sampling techniques and `MinMaxScaler` gave the best F1-score of 0.88 and one of the best ROC AUC. The model with the best ROC AUC is `LinearSVC` with `ROS (Random Over Sampling)` and a `StandardScaler` with a value of 0.937 (row 14 in the results table).

This shows that the classification problem is a non-linear one, as LinearSVC and Logistic Regression both struggle to produce a high enough F1-score, while Random Forest did produce a high F1-score, which is known for handling non-linear relationships well.

The highest performing `Logistic Regression` model is without any scaling or sampling techniques, with an F1-score of 0.67 and a ROC AUC of 0.93.

The highest performing `LinearSVC` is with `Random Over Sampling (ROS)` and the data being scaled using the `StandardScaler`. These techniques increased the F1-score from 0.49 to 0.48 and the ROC AUC from 0.92 to 0.94.

In [18]:
#dataset 2
dbtable = pd.read_csv("creditcard.csv")
X = dbtable.drop("Class", axis=1).copy()
y = dbtable["Class"]
numerical_cols_actual = [col for col in X.columns if X[col].dtype in ['float64', 'int64', 'float32', 'int32']]
all_results2 = model_comparison(models_to_test, scalers, samplers, X, y, numerical_cols_actual, verbose=1)

Splitting data: test_size=0.2, random_state=42, stratify=y
Train shape: (227845, 30), Test shape: (56962, 30)
Test label distribution:
Class
0    0.99828
1    0.00172
Name: proportion, dtype: float64


=== Evaluating Model: Logistic Regression ===

=== Evaluating Model: Random Forest ===

=== Evaluating Model: Linear SVC ===
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Li

In [20]:
all_results2

Model          Scaler Sampler  Accuracy  Precision  \
0         Random Forest            None    None  0.999596   0.941176   
1         Random Forest  StandardScaler    None  0.999596   0.941176   
2         Random Forest    MinMaxScaler    None  0.999561   0.929412   
3         Random Forest  StandardScaler   SMOTE  0.999491   0.870968   
4         Random Forest  StandardScaler     ROS  0.999526   0.949367   
5         Random Forest            None     ROS  0.999526   0.949367   
6         Random Forest    MinMaxScaler     ROS  0.999526   0.949367   
7         Random Forest    MinMaxScaler   SMOTE  0.999456   0.852632   
8         Random Forest            None   SMOTE  0.999421   0.835052   
9   Logistic Regression            None    None  0.999157   0.828947   
10  Logistic Regression  StandardScaler    None  0.999140   0.826667   
11           Linear SVC            None    None  0.999122   0.807692   
12           Linear SVC    MinMaxScaler    None  0.999087   0.828571   
13           Linear SVC  StandardScaler    None  0.999087   0.828571   
14  Logistic Regression    MinMaxScaler    None  0.998894   0.786885   
15  Logistic Regression    MinMaxScaler     RUS  0.996787   0.334630   
16  Logistic Regression            None   SMOTE  0.988308   0.118280   
17           Linear SVC    MinMaxScaler     RUS  0.985587   0.097887   
18           Linear SVC            None   SMOTE  0.982146   0.081130   
19           Linear SVC    MinMaxScaler     ROS  0.979548   0.072173   
20           Linear SVC  StandardScaler     ROS  0.979478   0.071942   
21           Linear SVC            None     ROS  0.979618   0.071716   
22  Logistic Regression    MinMaxScaler     ROS  0.977353   0.065598   
23           Linear SVC  StandardScaler   SMOTE  0.977283   0.064774   
24  Logistic Regression    MinMaxScaler   SMOTE  0.976651   0.063121   
25           Linear SVC    MinMaxScaler   SMOTE  0.976528   0.062809   
26  Logistic Regression  StandardScaler     ROS  0.975668   0.061308   
27  Logistic Regression  StandardScaler   SMOTE  0.974106   0.057803   
28  Logistic Regression            None     ROS  0.969313   0.049180   
29        Random Forest    MinMaxScaler     RUS  0.964116   0.042333   
30        Random Forest  StandardScaler     RUS  0.964116   0.042333   
31        Random Forest            None     RUS  0.964099   0.042313   
32           Linear SVC            None     RUS  0.962326   0.040395   
33  Logistic Regression  StandardScaler     RUS  0.960219   0.038330   
34           Linear SVC  StandardScaler     RUS  0.959412   0.037594   
35  Logistic Regression            None     RUS  0.957674   0.036101   

      Recall  F1 Score   ROC AUC  
0   0.816327  0.874317  0.963027  
1   0.816327  0.874317  0.963027  
2   0.806122  0.863388  0.963039  
3   0.826531  0.848168  0.968451  
4   0.765306  0.847458  0.962822  
5   0.765306  0.847458  0.962821  
6   0.765306  0.847458  0.962804  
7   0.826531  0.839378  0.963982  
8   0.826531  0.830769  0.964423  
9   0.642857  0.724138  0.948542  
10  0.632653  0.716763  0.960549  
11  0.642857  0.715909  0.942616  
12  0.591837  0.690476  0.962518  
13  0.591837  0.690476  0.943127  
14  0.489796  0.603774  0.979143  
15  0.877551  0.484507  0.977924  
16  0.897959  0.209026  0.976047  
17  0.897959  0.176530  0.978518  
18  0.908163  0.148954  0.971696  
19  0.918367  0.133829  0.975129  
20  0.918367  0.133432  0.974638  
21  0.908163  0.132935  0.971147  
22  0.918367  0.122449  0.974866  
23  0.908163  0.120924  0.974963  
24  0.908163  0.118037  0.975251  
25  0.908163  0.117492  0.975905  
26  0.918367  0.114943  0.972004  
27  0.918367  0.108761  0.970843  
28  0.918367  0.093361  0.972548  
29  0.918367  0.080935  0.977698  
30  0.918367  0.080935  0.977697  
31  0.918367  0.080899  0.977698  
32  0.918367  0.077386  0.974489  
33  0.918367  0.073590  0.976126  
34  0.918367  0.072231  0.974876  
35  0.918367  0.069471  0.975351

For Dataset 2, the highest performer was also the baseline `Random Forest` with no sampling techniques or data scaling applied to it. The baseline gave an F1-score of 0.82 with a ROC AUC of 0.96. The model that gave the highest ROC AUC however, is the Logistic Regression with no sampling techniques but the data is scaled using the `MinMaxScaler` with a ROC AUC  0.98 (row with the index of 14)

The highest performing `Logistic Regression` is also the baseline with no sampling techniques or data scaling applied to it. This gave an F1-score of 0.72 and a ROC AUC of 0.95.

The highest performing `LinearSVC` is also the baseline. It gave an F1-score of 0.72 and a ROC AUC of 0.94.

In [17]:
#dataset 1
df = pd.read_csv("creditcard_2023.csv")

X = df.drop(['id', 'Class'], axis = 1)
y = df['Class']
numerical_cols_actual = [col for col in X.columns if X[col].dtype in ['float64', 'int64', 'float32', 'int32']]
all_results1 = model_comparison(models_to_test, scalers, samplers, X, y, numerical_cols_actual, verbose=1)

Splitting data: test_size=0.2, random_state=42, stratify=y
Train shape: (454904, 29), Test shape: (113726, 29)
Test label distribution:
Class
1    0.5
0    0.5
Name: proportion, dtype: float64


=== Evaluating Model: Logistic Regression ===

=== Evaluating Model: Random Forest ===

=== Evaluating Model: Linear SVC ===
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SV

In [19]:
all_results1

Model          Scaler Sampler  Accuracy  Precision  \
0         Random Forest            None    None  0.999833   0.999666   
1         Random Forest            None   SMOTE  0.999833   0.999666   
2         Random Forest            None     ROS  0.999833   0.999666   
3         Random Forest  StandardScaler    None  0.999833   0.999666   
4         Random Forest  StandardScaler   SMOTE  0.999833   0.999666   
5         Random Forest  StandardScaler     ROS  0.999833   0.999666   
6         Random Forest    MinMaxScaler    None  0.999833   0.999666   
7         Random Forest    MinMaxScaler   SMOTE  0.999833   0.999666   
8         Random Forest    MinMaxScaler     ROS  0.999833   0.999666   
9         Random Forest    MinMaxScaler     RUS  0.999807   0.999613   
10        Random Forest            None     RUS  0.999807   0.999613   
11        Random Forest  StandardScaler     RUS  0.999807   0.999613   
12  Logistic Regression  StandardScaler    None  0.964977   0.977170   
13  Logistic Regression  StandardScaler   SMOTE  0.964977   0.977170   
14  Logistic Regression  StandardScaler     RUS  0.964977   0.977170   
15  Logistic Regression  StandardScaler     ROS  0.964977   0.977170   
16  Logistic Regression            None    None  0.964810   0.976956   
17  Logistic Regression            None   SMOTE  0.964810   0.976956   
18  Logistic Regression            None     ROS  0.964810   0.976956   
19  Logistic Regression            None     RUS  0.964327   0.977917   
20           Linear SVC  StandardScaler    None  0.963333   0.980153   
21           Linear SVC  StandardScaler   SMOTE  0.963333   0.980153   
22           Linear SVC  StandardScaler     ROS  0.963333   0.980153   
23           Linear SVC  StandardScaler     RUS  0.963333   0.980153   
24           Linear SVC            None    None  0.963210   0.980271   
25           Linear SVC            None   SMOTE  0.963210   0.980271   
26           Linear SVC            None     ROS  0.963210   0.980271   
27           Linear SVC            None     RUS  0.963210   0.980271   
28           Linear SVC    MinMaxScaler    None  0.962348   0.980358   
29           Linear SVC    MinMaxScaler   SMOTE  0.962348   0.980358   
30           Linear SVC    MinMaxScaler     ROS  0.962348   0.980358   
31           Linear SVC    MinMaxScaler     RUS  0.962339   0.980341   
32  Logistic Regression    MinMaxScaler     RUS  0.961618   0.980769   
33  Logistic Regression    MinMaxScaler    None  0.961618   0.980769   
34  Logistic Regression    MinMaxScaler   SMOTE  0.961618   0.980769   
35  Logistic Regression    MinMaxScaler     ROS  0.961618   0.980769   

      Recall  F1 Score   ROC AUC  
0   1.000000  0.999833  0.999990  
1   1.000000  0.999833  0.999990  
2   1.000000  0.999833  0.999990  
3   1.000000  0.999833  0.999990  
4   1.000000  0.999833  0.999990  
5   1.000000  0.999833  0.999990  
6   1.000000  0.999833  0.999990  
7   1.000000  0.999833  0.999990  
8   1.000000  0.999833  0.999990  
9   1.000000  0.999807  0.999990  
10  1.000000  0.999807  0.999990  
11  1.000000  0.999807  0.999990  
12  0.952201  0.964524  0.993500  
13  0.952201  0.964524  0.993500  
14  0.952201  0.964524  0.993500  
15  0.952201  0.964524  0.993500  
16  0.952078  0.964356  0.993478  
17  0.952078  0.964356  0.993478  
18  0.952078  0.964356  0.993478  
19  0.950108  0.963812  0.993377  
20  0.945817  0.962679  0.993314  
21  0.945817  0.962679  0.993314  
22  0.945817  0.962679  0.993314  
23  0.945817  0.962679  0.993314  
24  0.945448  0.962545  0.993309  
25  0.945448  0.962545  0.993309  
26  0.945448  0.962545  0.993309  
27  0.945448  0.962545  0.993309  
28  0.943601  0.961629  0.993196  
29  0.943601  0.961629  0.993196  
30  0.943601  0.961629  0.993196  
31  0.943601  0.961620  0.993197  
32  0.941702  0.960838  0.992522  
33  0.941702  0.960838  0.992522  
34  0.941702  0.960838  0.992522  
35  0.941702  0.960838  0.992522

Dataset 1 seems synthetic or manipulated as it was balanced, which is unrealistic for datacard fraud. For this dataset, the highest performer is `Random Forest` with no sampling or scaling techniques, with an F1-score of 0.99 and a ROC AUC of 0.99.

The highest performing `Logistic Regression` is with no sampling technique but with the `StandardScaler`. This increased the F1-score from 0.964356 to 0.964524, a very small increase of 0.000168. The ROC AUC also increased very slightly.

The highest performing `LinearSVC` is also with no sampling technique but with the `StandardScaler`. It got an F1-score of 0.964 and a ROC AUC of 0.9934. Compared to the baseline `LinearSVC` with no techniques, it also had very minute changes.

## Discussion
The experimental investigation across three distinct datasets provided data on the comparative performance of Random Forest, Logistic Regression, and Linear SVC for credit card fraud detection. The impact of feature scaling and class imbalance sampling techniques was also assessed. The primary evaluation metric was the F1-score, due to its relevance in imbalanced classification, with the ROC AUC score providing a complementary measure of model discriminative ability.

On Dataset 3, which presented challenges for linear models, Random Forest achieved the highest F1-score. The configuration using `MinMaxScaler` without sampling yielded an `F1-score` of approximately `0.879`, while the baseline Random Forest (no preprocessing) also performed well with an F1-score of 0.877. This suggests Random Forest's capacity to model the data effectively. In contrast, Logistic Regression and Linear SVC produced lower F1-scores on this dataset. Logistic Regression's best F1-score was 0.671 without preprocessing, and Linear SVC's was approximately 0.559 with StandardScaler and RandomOverSampler. These results indicate that a linear function may not adequately approximate the decision boundary for this dataset. Sampling techniques like SMOTE and RUS, when applied to Random Forest, generally increased recall but decreased precision, leading to lower F1-scores. This suggests that the introduced or remaining samples might have adversely affected the model's precision-recall balance.

For Dataset 2, Random Forest again achieved the highest F1-score (0.874) in its baseline configuration. This dataset also favored the non-linear modeling capabilities of Random Forest. Linear models showed some response to preprocessing; for instance, Logistic Regression achieved its highest ROC AUC with MinMaxScaler, though its F1-score did not surpass its baseline. Linear SVC performed best with its baseline setup. Similar to Dataset 3, common sampling techniques generally did not improve, and often reduced, the F1-scores for all models on Dataset 2, primarily through a negative impact on precision.

Dataset 1 appeared to be either balanced or synthetically generated, leading to high performance across all models. Random Forest achieved high F1-scores with minimal preprocessing. Logistic Regression and Linear SVC also performed well, with minor, if any, improvements observed when StandardScaler was applied. Given the dataset's characteristics, sampling techniques offered no discernible benefits.

Across all datasets, several patterns were observed. Random Forest consistently produced high F1-scores, often with no specific scaling or sampling. This indicates its ability to handle complex data structures. Linear models, Logistic Regression and Linear SVC, were more influenced by dataset characteristics. Their performance was lower than Random Forest on Dataset 3. Scaling sometimes offered slight benefits to their ROC AUC scores or specific F1 components but did not enable them to match Random Forest's F1-scores where non-linearity was a factor. The investigated sampling techniques showed mixed and often unfavorable results for F1-score improvement. While recall for the minority class was often increased, this typically occurred at the expense of precision. Scalers had minimal impact on Random Forest, as anticipated, while their effect on linear models was inconsistent and generally modest for the F1-score.

## Conclusion
In summary, this empirical evaluation indicates that Random Forest generally achieved the highest F1-scores among the three algorithms tested for credit card fraud detection, particularly on the imbalanced datasets with apparent non-linear characteristics (Datasets 2 and 3). Its performance often peaked with minimal preprocessing. Logistic Regression and Linear SVC, while offering different modeling approaches, did not match Random Forest's F1-scores on the more complex datasets. Feature scaling provided inconsistent and generally minor improvements to F1-scores for these linear models. The application of the selected sampling techniques did not consistently lead to better F1-scores; often, recall for the minority class increased while precision decreased, resulting in a net negative or negligible impact on the F1-score. These findings suggest that for the datasets and configurations tested, an algorithm like Random Forest, capable of addressing non-linear relationships, was more effective than the linear models, and that standard preprocessing steps like scaling and sampling require careful, dataset-specific validation to ensure they provide a net benefit to the chosen performance metric.